### DISCUSSION TITLE
关于部分用户不显示控制面板的原因及解决方法

### DISCUSSION THREAD

**Discussion by andywang425** [andywang425/BLTH#453] [Development]
目前一部分用户在QQ频道反馈没有控制面板，根据我这段时间和其中一位用户的交流，我发现问题似乎出在这里：

https://github.com/andywang425/BLTH/blob/dc323ccaeffc1376197fb910e2b6bade360c21dd/src/main.ts#L16

因为之前有个版本，这行代码之前有一行日志输出`logger.log('document.readyState', document.readyState)`，我能在他的控制台截图里找到这条且`document.readyState`为`loading`，后面就没有任何日志了。后来我把这行日志输出放到了下面，该用户就一条BLTH日志输出都没有了。

`isTargetFrame`函数根据`window.BilibiliLive`来判断当前是否是目前frame。`window.BilibiliLive`的定义写在`document.head`里的一个script代码块里，且是最初就存在于html里的。
![image](https://github.com/andywang425/BLTH/assets/56548456/a7bc6ab1-6419-4084-b753-d071747b7e4c)

所以我推断：由于脚本被注入得太早，`window.BilibiliLive`还不存在，那时候`document.head`都没出现。目前脚本的`@run-at`是`document-start`，我查阅了Tampermonkey和Violentmonkey对`document-start`的解释，确实也没说那时候`document.head`就一定存在了。

我看了下Blibili-Evolved的代码，有一段等待`document.head`出现的代码：

https://github.com/the1812/Bilibili-Evolved/blob/c14a2332244f88f8c749305b7279601f21833456/src/client/init.ts#L26-L29

我在想是不是BLTH是不是也应该这么做。欢迎各位大佬来讨论原因和解决方法。

**Comment by andywang425** [comment]
@ADJazzzz 哥们有空来看下不

**Comment by ADJazzzz** [comment]
我试着在你所说的那里打断点，`window.BilibiliLive`返回值是`undefined`，判断为`true`
![1](https://github.com/andywang425/BLTH/assets/68748165/19ac751f-9511-4c6b-8442-a69845f226cb)
所以说这个判断是不是有问题

**Reply by andywang425** [reply]
那是指当前作用域下`this`为`undefined`，因为箭头函数是没有this的。

**Reply by ADJazzzz** [reply]
`@run-at`换到`document-body`会不会好一些

**Reply by andywang425** [reply]
对Tampermonkey来说这办法可以，但暴力猴不支持 https://violentmonkey.github.io/api/metadata-block/#run-at

**Reply by ADJazzzz** [reply]
要兼容的话目前我没想到什么好的办法，只能试试Blibili-Evolved的方法了

**Comment by andywang425** [comment]
今天写的一段测试代码：
```js
// ==UserScript==
// @name         test document-start
// @namespace    http://tampermonkey.net/
// @version      0.1
// @description  try to take over the world!
// @author       You
// @match        https://live.bilibili.com/22474988
// @grant        none
// @run-at       document-start
// ==/UserScript==

(function() {
    console.log('document.head',document.head.children,window.BilibiliLive)
})();
```

勾选控制台的网络选项卡中的【停用缓存】，结果：
![image](https://github.com/andywang425/BLTH/assets/56548456/157939e2-f8bc-42f9-89ca-2d113a530f3a)

此时head不为空，但是`window.BilibiliLive`依旧是`undefined`。

**Comment by andywang425** [comment]
找到了一种很直接的解决方案，~稳定有效~：
```js
const isTargetFrame = (): boolean => {
  if (unsafeWindow.document.head.innerHTML.includes('BilibiliLive')) {
    return true
  } else {
    return false
  }
}
```

**Reply by andywang425** [reply]
接下来打算再设计一套逻辑去管理各个模块的运行时机，有些模块需要尽早（拦截日志上报等），有些则晚点比较好（自动切换画质）。

**Reply by ADJazzzz** [reply]
自动切换画质开启会转圈，这个能解决吗？估计是执行太早播放器没反应过来

**Reply by andywang425** [reply]
对，慢点就好了。

**Comment by andywang425** [comment]
这个问题还没有解决。最近又和一些用户交流了一下，不显示控制面板还是因为脚本被注入得太早，具体来说有两个原因：

- `document.head`为`null`
- `document.head`存在但是只有一部分节点（比如`meta`，`link`），没有`script`节点，因此脚本检测不到`BilibiliLive`

**Comment by andywang425** [comment]
还有一个发现就是，violentmonkey是支持`document-body`的（通过看源代码），但是文档里没写。
